store/src/sql/event.rs

```diff
@@ -552,11 +552,18 @@ impl ceramic_api::AccessModelStore for SqliteEventStore {
         end: &EventId,
         offset: usize,
         limit: usize,
-    ) -> anyhow::Result<Vec<(EventId, Vec<u8>)>> {
-        let res = self
-            .range_with_values_int(start, end, offset, limit)
-            .await?;
-        Ok(res.collect())
+    ) -> anyhow::Result<Vec<(Cid, Vec<u8>)>> {
+        self.range_with_values_int(start, end, offset, limit)
+            .await?
+            .map(|(event_id, value)| {
+                Ok((
+                    event_id
+                        .cid()
+                        .ok_or_else(|| anyhow!("EventId does not have an event CID"))?,
```

Comment:
okay, updated this to handle if the CID is missing.  It compiles, but I'm not totally sure it's correct (or if it is I'm not entirely sure *why* it's correct).  It was actually kind of surprising to me that this compiled.

Previously, the map call returned a Map iterator structure where the values inside were a tuple of `(Cid, Vec<u8>)`.  Now the values returned in the Map are Results of that tuple (ie `Result<(Cid, Vec<ui>)>`).

I would have expected calling `collect()` on an `Iterator<Result<T>>` to return a `Vec<Result<T>>`.  But it looks like instead it is returning a `Result<Vec<T>>`?  Is it just taking the first non-ok Result and promoting that up to the top-level error?  That's what I want, but it wasn't/isn't clear to me that that's what `collect()` is doing.  How can I learn how this works?

Reply:
oh wow, okay I found this page that indicates that collect is indeed doing what I want: https://doc.rust-lang.org/rust-by-example/error/iter_result.html#fail-the-entire-operation-with-collect

This is actually quite nice, but feels unintuitive to me.  Any advice on how I can build an intuition (or get my IDE to tell me) around these types of behaviors in Rust would be appreciated

Reply:
tbh, I don't think it is super intuitive and I'm kind of surprised it works. I didn't think you could use `?` in closures so I need to review what's happening as well. The fact that [result implements FromIterator](https://doc.rust-lang.org/src/core/result.rs.html#1934) is just syntax sugar and imo kind of odd, but it is nice to avoid all the `transpose` and `map_err` business. 

I tend to use `filter_map` the other suggestions in that doc to log every error when it makes sense. The issue with the "abort immediately" approach here is that it will *never* succeed. The event data is garbage, so someone is going to have to manually fix it... the `filter_map` approach would let it return something, but we shouldn't be storing this data in the first place, and if we do have it, someone or something is going to need to intervene.

Reply:
My intuition for it is that `collect` can construct various types from iterators. As its common to want to construct a vector from an iterator of results and fail on the first one  the Result type implements FromIteraor as David mentions. That is part one of this code, you can collect an iterator of results into a result of a collection failing on the first error.


Part two is that we explicitly return a Result from the `map` closure. There is nothing special about closures or the `?` syntax so using it within a closure is reasonable to me. To _help_ the closure understand we want to return a result we make the `Ok` variant explicit and then use `?` where needed.



Rust idoms aside why do we want to only log this error? Seems like if we have bad data we should error the request? Maybe I am missing something.

Reply:
Yeah, good call that ? is fine if you return a result. 

So I guess my argument for not an error is that this is only returning data we already stored, not preventing the data from being stored in the first place. So it'll never possible to stop erroring until someone goes and deletes something. IMO skipping bad data in the range would be reasonable instead of just failing until someone intervenes (or we could just delete the data because we don't expect it in our table) but it shouldn't happen.